marimo/_runtime/copy.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2024 Marimo. All rights reserved.
+from __future__ import annotations
+
+import inspect
+import sys
+import weakref
+from copy import copy
+from typing import (
+    Any,
+    Callable,
+    Generic,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+T = TypeVar("T")
+if sys.version_info < (3, 9):
+    # Future does not seem to work for this in CI.
+    Ref = Union["weakref.ReferenceType[T]", Callable[[], T]]
+else:
+    Ref = Union[weakref.ReferenceType[T], Callable[[], T]]
+
+
+class CloneError(Exception):
+    """Thrown when strict execution fail to deep copy or clone."""
+
+
+class RoError(Exception):
+    """Thrown when attempting to modify a read-only object."""
+
+
+class _Copy(Generic[T]):
+    """Base wrapper class for strict execution."""
+
+    __ref__: Ref[T]
+
+
+class ZeroCopy(_Copy[T]):
+    """Wrapper class for strict execution (stops deepcopy)."""
+
+
+class ShallowCopy(_Copy[T]):
+    """Wrapper class for strict execution (does copy over deepcopy)."""
+
+
+def _ro_fail() -> None:
+    raise RoError(
+        "Weakly copied objects are directly read only."
+        " Modification is not encouraged, but is"
+        " possible by utilizing the marimo.unwrap_copy() function."
+    )
+
+
+def shadow_wrap(ref_cls: Type[_Copy[T]], base: T) -> T:
```

Comment:
Is the point of this to try to prevent mutations of attributes? My understanding is that re-assignment will fail, but mutations by calling methods on attributes will still work.

Reply:
This is the closest I could get to a "move"

Idea is we check the variable type for ZeroCopy or ShallowCopy to determine clone behavior. Preventing direct mutation is a needed consequence of exposing all the attributes.

We could also not expose the attributes and always require an unwrap or call, and not do this "shadowing"

Reply:
Zero copy could just put the ID in a special dictionary. Much less complicated, but does require some trickery 

I think this is a better idea, just type checking was my first idea. Thoughts?

Reply:
Type checking does have the consequence of being conscious of your global variables though, and the RO forces you to unwrap and makes it clear state change is happening.

Particular to numpy, there is `.view()` which can called.

Reply:
> This is the closest I could get to a "move"

As in `std::move` from C++? I'm not sure I see the connection, maybe I'm missing something obvious

> Zero copy could just put the ID in a special dictionary. Much less complicated, but does require some trickery

And then get the original object in subsequent cells? That might be simpler, and would fix the issue with associative operators that you mentioned. It would also be compatible with any code that relied on checking the `type` of a zero copied object, whereas wrapping doesn't appear to be:

<img width="1158" alt="image" src="https://github.com/marimo-team/marimo/assets/1994308/d2f4bd22-96ba-4ec3-8dd6-dfb99e1a1000">

Could a special dictionary be used for `ShallowCopy`, too?

Reply:
> Type checking does have the consequence of being conscious of your global variables though, and the RO forces you to unwrap and makes it clear state change is happening.

By the time you're using strict and opting into things like `zero_copy` — I imagine you're already being pretty conscious of your global variable usage. 

> the RO forces you to unwrap and makes it clear state change is happening

That's true.

I would opt for a special dictionary because it's simpler and unlike type checking doesn't have gotchas when interacting with the "wrapped" object. But I'll leave it up to you to determine the benefits of typing outweigh its costs.

Reply:
Implementation wise, there would just be a special dictionary for zero and shallow copy, that the helper functions would just interact with, and then a check would be made from the global objects. Invalidating IDs on rerun might be tricky. Also, Python reuses IDs. But if not going down the id route, then there might be some manual gc that would have to happen 

I don't think it would be hard to put in but maybe a few edge cases- so I'm inclined to leave it as is for now

Reply:
> Implementation wise, there would just be a special dictionary for zero and shallow copy, that the helper functions would just interact with, and then a check would be made from the global objects. Invalidating IDs on rerun might be tricky. Also, Python reuses IDs. But if not going down the id route, then there might be some manual gc that would have to happen

Yea. That makes sense to me. I had to solve related edge cases for our registry of UI elements, during which process I learned that Python reuses IDs.

> I don't think it would be hard to put in but maybe a few edge cases- so I'm inclined to leave it as is for now

Ok sounds good.